### PR TITLE
[EVM] Improve BYTE instruction selection

### PR DESCRIPTION
**Pull request by PavelKopyl**
This replaces 'AND (SRL imm, v), 0xFF' -> 'BYTE (imm / 8), v',  where imm % 8 == 0.

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMISelDAGToDAG.cpp [hunk 1]
```diff
@@ -85,6 +85,32 @@ void EVMDAGToDAGISel::Select(SDNode *Node) {
     ReplaceNode(Node, CallResults);
     return;
   }
+  case ISD::AND: {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMISelDAGToDAG.cpp]: It seems we can tablegen it: https://chatgpt.com/share/688487a0-3554-8006-9579-ab02d2a890e4

**Review by akiramenai — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/EVM/EVMISelDAGToDAG.cpp [hunk 2]
```diff
@@ -85,7 +85,6 @@ void EVMDAGToDAGISel::Select(SDNode *Node) {
     ReplaceNode(Node, CallResults);
     return;
   }
-
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMISelDAGToDAG.cpp]: Please, revert this change

File: llvm/lib/Target/EVM/EVMInstrInfo.td [hunk 3]
```diff
@@ -156,6 +162,11 @@ def truncstorei8_heap : PatFrag<(ops node:$val, node:$ptr),
   let MemoryVT = i8;
 }
 
+// Immediate predicate: imm divisible by 8 and within [0, 31*8].
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.td]: Why not to put this pat next to the `byte_shift`?

File: llvm/lib/Target/EVM/EVMInstrInfo.td [hunk 4]
```diff
@@ -351,6 +362,8 @@ defm BYTE
       [(set GPR:$dst, (int_evm_byte GPR:$idx, GPR:$val))],
       "BYTE", "$dst, $idx, $val", 0x1a, 3>;
 
+def : Pat<(and(srl i256:$val, byte_shift_imm:$sh), (i256 255)),
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.td]: nit: Could you add spaces between outer instruction names and their first argument?

File: llvm/test/CodeGen/EVM/folding-to-byte.ll [hunk 5]
```diff
@@ -0,0 +1,17 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s | FileCheck %s
+target datalayout = "E-p:256:256-i256:256:256-S256-a:256:256"
+target triple = "evm-unknown-unknown"
+
+define i256 @test(i256 %0) {
```
Comment (akiramenai) [inline llvm/test/CodeGen/EVM/folding-to-byte.ll]: Could you please add more test cases including negative?

**Review by akiramenai — CHANGES_REQUESTED** [review]

**Review by akiramenai — APPROVED** [review]
LGTM

File: llvm/lib/Target/EVM/EVMStackModel.cpp [hunk 6]
```diff
@@ -91,6 +91,8 @@ Stack EVMStackModel::getSlotsForInstructionUses(const MachineInstr &MI) const {
       In.push_back(getStackSlot(MO));
     else if (MO.isMCSymbol())
       In.push_back(getSymbolSlot(MO.getMCSymbol(), &MI));
+    else if (MO.isCImm())
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMStackModel.cpp]: ~~This shouldn't be part of this PR, right?~~ Scratch that, I see why it is needed.
Is there a reason why for BYTE we embedded immediate inside instruction, whereas for other instructions we are not doing that? Maybe this change (to move immediates inside instructions) should be part of larger refactoring if we want to do that.

File: llvm/lib/Target/EVM/EVMInstrInfo.td (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.td]: We used to place Pat and 'XFrom' transformations in different sections, but yes it doesn't actually matter.

File: llvm/lib/Target/EVM/EVMStackModel.cpp (hunk 6 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMStackModel.cpp]: You are right, this is not required here, we can use CONST_I256. On the other hand it is a bit easier to be able to use immediate operands in instructions. Right, removing of CONST_I256 would be interesting to try (I least, this may speed up compilation). Should I CONST_I256 in this patch and try everything else in a separate patch?

Comment (vladimirradosavljevic) [reply llvm/lib/Target/EVM/EVMStackModel.cpp]: Let's do it in a separate patch.
Yeah, for sure that will improve compilation speed, but in some cases it could change heuristics in common passes, so we might see some changes in numbers.

**Review by vladimirradosavljevic — APPROVED** [review]
LGTM. Thanks!

File: llvm/lib/Target/EVM/EVMStackModel.cpp (hunk 6 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMStackModel.cpp]: Good, fixed to use CONST_I256.
